src/mscorlib/src/System/String.cs

```diff
@@ -465,88 +458,122 @@ private unsafe static bool StartsWithOrdinalHelper(String str, String startsWith
-                        return order;
-                    }
-                    Contract.Assert( *(a+1) != *(b+1), "This byte must be different if we reach here!");
-                    return ((int)*(a+1) - (int)*(b+1));                    
+                    if (*(int*)a != *(int*)b) goto DiffOffset0;
+                    if (*(int*)(a + 2) != *(int*)(b + 2)) goto DiffOffset2;
+                    if (*(int*)(a + 4) != *(int*)(b + 4)) goto DiffOffset4;
+                    if (*(int*)(a + 6) != *(int*)(b + 6)) goto DiffOffset6;
+                    if (*(int*)(a + 8) != *(int*)(b + 8)) goto DiffOffset8;
+                    length -= 10; a += 10; b += 10; 
                 }
+#endif
 
-                // now go back to slower code path and do comparison on 4 bytes at a time.
+                // Fallback loop:
+                // go back to slower code path and do comparison on 4 bytes at a time.
                 // This depends on the fact that the String objects are
                 // always zero terminated and that the terminating zero is not included
                 // in the length. For odd string sizes, the last compare will include
                 // the zero terminator.
-                while (length > 0) {
-                    if (*(int*)a != *(int*)b) {
-                        break;
-                    }
+                while (length > 0)
+                {
+                    if (*(int*)a != *(int*)b) goto DiffNextInt;
                     length -= 2;
                     a += 2; 
                     b += 2; 
                 }
 
-                if( length > 0) { 
-                    int c;
-                    // found a different int on above loop
-                    if ( (c = (int)*a - (int)*b) != 0) {
-                        return c;
-                    }
-                    Contract.Assert( *(a+1) != *(b+1), "This byte must be different if we reach here!");
-                    return ((int)*(a+1) - (int)*(b+1));                                        
-                }
-
                 // At this point, we have compared all the characters in at least one string.
                 // The longer string will be larger.
                 return strA.Length - strB.Length;
+                
+#if BIT64
+                DiffOffset8: a += 4; b += 4;
+                DiffOffset4: a += 4; b += 4;
+#else
+                DiffOffset8: a += 2; b += 2;
+                DiffOffset6: a += 2; b += 2;
+                DiffOffset4: a += 2; b += 2;
+                DiffOffset2: a += 2; b += 2;
+#endif
+                
+                DiffOffset0:
+                // If we reached here, we already see a difference in the unrolled loop above
+#if BIT64
+                if (*(int*)a == *(int*)b)
+                {
+                    a += 2; b += 2;
+                }
+#endif
+                
+                DiffNextInt:
+                int order;
+                if ( (order = (int)*a - (int)*b) != 0) {
```

Comment:
Nope; don't know why I kept it while changing the code that subtracted `m_firstChars`. Will change it to the simpler form.